File: Data Structures & Algorithms/string-encode-and-decode/submission-4.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        res = ""
        for word in strs:
            res += str(len(word));
            res += "_" + word;
        return res;

    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        n = len(s)
        while i < n:
            j = i
            while j < n and s[j] != '_':
                j += 1
            length = int(s[i:j])
            j += 1
            res.append(s[j: j + length])
            i = j + length
        return res

File: Data Structures & Algorithms/anagram-groups/submission-4.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> groups;
        for (const auto& s: strs) {
            vector<int> freq(26, 0);
            for (char c: s) {
                ++freq[c - 'a'];
            }
            string key = "";
            for (int n: freq) {
                key += '#' + n;
            }
            groups[key].push_back(s);
        }
        vector<vector<string>> res;
        for (const auto&[_, group]: groups) {
            res.push_back(group);
        }
        return res;
    }
};
